### ISSUE TITLE
Replace prereq and post* with pre and post middleware?

### ISSUE THREAD

**Issue by Zertz** [florianholzapfel/express-restify-mongoose#157]
I've been hacking away at a fairly complex piece of code within an API endpoint and I've come to realize that `prereq` doesn't do anything that `middleware` can't.

The few things `prereq` does, `middleware` does better, plus you get the familiarity and flexibility. Same with postCreate, postUpdate and postDelete.

Here's my suggestion:

* Remove `prereq`
* Rename `middleware` to `preMiddleware`:

```javascript
preMidleware: [function (req, res, next) {
  ...
}, ...]
```

* Remove all 3 `post*` methods and add `postMiddleware`:

```javascript
postMidleware: [function (req, res, next, result) {
  ...
}, ...]
```

*`result` is passed as an extra parameter to avoid polluting the `req` object.*

Unless there are any downsides or loss of functionality that I'm not seeing, I'm very much willing to get this code up and running! This would be part of a 2.0 release.

**Comment by arathael** [comment]
I concur, for the sake of clarity about how these options work. It is easier for to work with familiar concepts as express middleware than custom API methods. I have always struggled with those postOp signatures.

**Comment by Zertz** [comment]
It's not set in stone yet, but the flow will probably be `preMiddleware` > `access` > query > `outputFn` > `postMiddleware` with everything being optional.

**Comment by arathael** [comment]
I would rather prefer `postMiddleware` to execute before outputFn. My rationale is this:

If I have to do some transactional output, let's say my last use case:
```http
PUT /Events
{
  "status": "started"
}
```
When I start an `Event` my app uses a 3rd party API to create a custom subscription, so I should return `200` if everything goes ok. The `postMiddleware` allows me to do the API calling and failing the whole, while reserve `outputFn` to really just format my output json.

I know I can do this on the `preMiddleware` but I need the db validation middleware to run before doing the call.

What do you think?

**Comment by Zertz** [comment]
@arathael It does make more sense to have `postMiddleware` first. If someone really needs to do stuff after `outputFn`, then it's possible to call `next()` from there and implement some middleware after the API routes.

**Comment by arathael** [comment]
Or, as in my case, when the post processing logic has nothing to do with the http layer (which this module is an aid for), you could always do some Mongoose post save middleware, and really don't support that kind of features in your lib.

**Comment by Zertz** [comment]
I'm beginning to wonder if it's even necessary to have `postMiddleware` since that's essentially what `outputFn` is.

**Comment by Zertz** [comment]
@arathael See PR #154

**Comment by Zertz** [comment]
Implemented in #154 

`access` -> `preMiddleware` -> query -> `postCreate/Read/Update/Delete` -> `output` -> `postProcess`
